### ISSUE TITLE
Some suggestions regarding the new `jinx-correct` UI

### ISSUE THREAD

**Issue by gusbrs** [minad/jinx#43]
Hi @minad I took the new UI for a spin on some actual documents, and it's really good, thank you! I do have some suggestions though.

I particularly considered the case of whole buffer spell checking, that is calling `jinx-correct` with an argument. The scenario is the following: a large LaTeX document, 20k+ lines of 78 cols hard wrapped content, generating a 500+ page pdf, document has one main language with snippets of 3-4 other languages, but I wouldn't want to add those to `jinx-languages` to avoid false negatives. A call to `C-u M-x jinx-correct` gives me 11k+ misspellings, almost all of them false positives.

In such a document sometimes I'm faced with a situation such as:

![Screenshot from 2023-04-19 09-50-57](https://user-images.githubusercontent.com/19410606/233080261-197af8d4-978e-482c-bf59-84e761c07911.png)

So that I have a long sequence of obvious skips. Skipping them one by one is not appealing, but it is somewhat hard to do otherwise. The attempt to `C-x o` move point and `C-x o` again, does not work at all, since `jinx-correct` is looping over the overlays, not working its way through buffer positions. The only real alternative is to `C-g`, move point, and call `C-u M-x jinx-correct` again. The problem here is that this will force check the whole buffer again, which in this case is a costly operation, taking around 10s. `(benchmark-run 10 (jinx--force-check-region (point-min) (point-max)))` returns `(107.914697209 787 62.143297797)`. To be fair, it is a great performance, `(benchmark-run 10 (flyspell-buffer))` takes `(346.755893513 130 10.590287985999993)` on the same file. But 10s is enough to break the workflow we discussed at https://github.com/minad/jinx/issues/9#issuecomment-1487593763 of interrupting `C-u M-x jinx-correct`, "do something", and continuing with `C-u M-x jinx-correct`. You think thrice before hitting `C-g`.

Also, interrupting with `C-g` and starting over with `C-u M-x jinx-correct` generates the overlays anew from point, so you loose the "original starting point" of your spell checking session, from the perspective of the overlays list.

That given, I have a couple of suggestions. The first one is pretty simple, but I think it is useful. Let `jinx-correct-next` and `jinx-correct-previous` receive numeric arguments, pass them on the throw, so that skipping multiple overlays is possible.

The second one is about the behavior of `C-g` and how we could improve reiterated calls to `C-u M-x jinx-correct`. Of course, the first time it is called, it has to check the whole buffer, there's no way around it. But on reiterated ones, it could be smarter and continue where it left from. I'm not sure of what the best take here, but I do have a concrete suggestion. When called with an argument and interrupted with a `C-g` `jinx-correct` could store in a variable the list of overlays it was operating on and the current overlay. `C-u M-x jinx-correct` could then "continue the session", without the need to regenerate the overlays again and without losing the "original starting point" of the session. A call to `C-u C-u M-x jinx-correct` could be made to ignore such variable and force the checking of the buffer again from point. WDYT?

Finally, on a different theme, I also noticed a weird behavior of `undo` when called during a `C-u M-x jinx-correct` call. Suppose you correct one word, regret it, and decide to undo it. If you press `C-/` you get the message "No further undo information", and indeed the operation is not carried out. Trying to `C-x o` then `C-/` is weirder, it ignores the last correction made by Jinx, and undoes the previous change.

**Comment by oantolin** [comment]
I don't mean to detract from the valid UX points you make, but you should definitely consider using both Portuguese and English dictionaries at the same time. I have a bunch of files that mix English and Spanish, and using the dictionaries simultaneously has been fantastic.

**Comment by minad** [comment]
@gusbrs 

> ...a large LaTeX document, 20k+ lines of 78 cols hard wrapped content, generating a 500+ page pdf, document has one main language with snippets of 3-4 other languages, but I wouldn't want to add those to jinx-languages to avoid false negatives.

If your document is only partially in another language I understand that you don't want to use multiple languages in `jinx-languages`. In this case I'd suggest a custom predicate added to `jinx--predicates` which filters out these blocks.

> That given, I have a couple of suggestions. The first one is pretty simple, but I think it is useful. Let jinx-correct-next and jinx-correct-previous receive numeric arguments, pass them on the throw, so that skipping multiple overlays is possible.

Sounds like a good idea. I will add this.

> The second one is about the behavior of C-g and how we could improve reiterated calls to C-u M-x jinx-correct. Of course, the first time it is called, it has to check the whole buffer, there's no way around it. But on reiterated ones, it could be smarter and continue where it left from. I'm not sure of what the best take here, but I do have a concrete suggestion. When called with an argument and interrupted with a C-g jinx-correct could store in a variable the list of overlays it was operating on and the current overlay.

Honestly, I am not interested in adding this complication.

> Finally, on a different theme, I also noticed a weird behavior of undo when called during a C-u M-x jinx-correct call.

Hmm, I could not reproduce this. Which Emacs version do you use? Can you give me a precise recipe?

**Comment by minad** [comment]
I added support for prefix arguments. Regarding your second point I think I could refactor the code somewhat such that the list of checked overlays is more flexible. The current mechanism is hard coded here:

https://github.com/minad/jinx/blob/f359e84776cb65fa9386b07070929bc8b975f80b/jinx.el#L718-L727

Please open a separate issue for the undo problem with more detail.

**Comment by minad** [comment]
Two more comments out of interest. 

@gusbrs Could you please check if spell checking itself is the expensive part here? I suspect that the 10s are mostly eaten up by font locking, on which we have to rely for obvious reasons.

@oantolin Did you try Jinx on math LaTeX documents? Jinx is unfortunately a bit lacking there since it is unable to filter out equation/align blocks. One could use `texmathp` as a predicate but from what I've seen it is quite inefficient. Jinx deliberately cannot use a lot of the existing machinery for performance reasons. For example I cannot even use the flyspell mode predicate since the Org implementation is O(buffer-size).

**Comment by gusbrs** [comment]
> If your document is only partially in another language I understand that you don't want to use multiple languages in `jinx-languages`. 

Yes, it is 95% in one language, and for this kind of task I really don't want to introduce the possibility of false negatives a second dictionary creates. Besides, adding 4-5 dictionaries would not be wise even if that was not the case. (@oantolin That's why I don't want to add English too. This is fine for my notes and I do it, but not for a formal document like this, for which I want more precise results).

> In this case I'd suggest a custom predicate added to `jinx--predicates` which filters out these blocks.

But, short of something like https://github.com/minad/jinx/issues/15, how to know the block is in a different language? You mean introduce some custom markup, and exclude that?

> Sounds like a good idea. I will add this.

Thank you.

> Honestly, I am not interested in adding this complication.

A pity. I was hoping for some way at least to avoid the rechecking of the whole buffer every time, it seems a waste of sorts, since the overlays are already in place.

> Regarding your second point I think I could refactor the code somewhat such that the list of checked overlays is more flexible or could be passed as argument even. This would make the checking code more easily tweakable for the user.

But I see you are already onto that.  Thanks.  :-)

> I could not reproduce this in a quick try. Which Emacs version do you use? Can you give me a precise recipe?

The Emacs 29 prerelease. I'll examine it in more detail and report a separate issue if I can find a precise recipe.

**Comment by gusbrs** [comment]
> Could you please check if spell checking itself is the expensive part here? I suspect that the 10s are mostly eaten up by font locking, on which we have to rely for obvious reasons.

That is possible, yes, but unavoidable anyway. But I'm not sure how to check that. Do you have some particular test in mind?

**Comment by oantolin** [comment]
I haven't tried it on LaTeX documents yet, @minad. I actually only decided to turn jinx on all the time yesterday.

**Comment by minad** [comment]
@oantolin I see. Please let me know how it goes. There are probably still many rough edges.

@gusbrs 

> But, short of something like https://github.com/minad/jinx/issues/15, how to know the block is in a different language? You mean introduce some custom markup, and exclude that?

Yes, custom markup or custom comments. It would be interesting to have some functionality to exclude blocks (scanning backward and forward to some begin/end comments). This goes in the direction of `ispell-skip-region-alist`, but I am concerned about the performance impact.

> A pity. I was hoping for some way at least to avoid the rechecking of the whole buffer every time, it seems a waste of sorts, since the overlays are already in place.

Well, but this opens a whole other set of issues to determine if the overlays are still up to date and so on. I want to avoid maintaining a separate list of overlays.

**Comment by gusbrs** [comment]
> Yes, custom markup or custom comments.

That might be a good idea. Thanks.

> Could you please check if spell checking itself is the expensive part here? I suspect that the 10s are mostly eaten up by font locking, on which we have to rely for obvious reasons.

Regarding that, some preliminaries. It seems I'm in large part responsible for the time it `jinx` is taking. If I disable my customizations (as in https://github.com/minad/jinx/issues/25#issuecomment-1511954696), the time falls to about 3s. It's still a gap in the workflow but, well, less than 10s.

Also, `(benchmark-run 10 (font-lock-ensure (point-min) (point-max)))` takes `(1.596758774 7 0.5862017130000083)` with personal customizations, and `(1.491363519 6 0.5119681749999927)` without them. Which, as far as I understand, means font-locking is not the culprit, but checking is. It also means I must be doing something wrong in my personal predicates, I'll have to do some profiling to see where it comes from.

**Comment by gusbrs** [comment]
> Okay, this means half of the time is taken by font locking and half of the time is taken by the actual checking.

No, that's the time of 10 runs, font-locking takes a fraction of the time. ~5% without modifications and with my customizations ~2%.

**Comment by minad** [comment]
Oh, I see. Maybe TeX font locking is very fast. When I tried `jinx-correct` in large Org buffers the story was different. Font locking seems more expensive there. Anyway this at least confirms that the Jinx is on the right track, since in normal mode Jinx only spell-checks the visible region and only after an idle time. In other words, performance of whole buffer checks was and is completely irrelevant to the main value of Jinx.

**Comment by gusbrs** [comment]
> This is just anecdotal evidence [...]

It certainly is.

> It just occurred to me, but maybe your benchmark is flawed? Even if you run `font-lock-ensure` ten times it will only apply the font locking _once_ since the font locking code keeps track of modifications.

It may be, I thought about that too. And indeed, repeated calls significantly changes the value. But this would affect the calls to `jinx--force-check-region`. So I guess, "less than anecdotal" is really more like it. :shrug: 

> Anyway this at least confirms that the Jinx is on the right track, since normally Jinx only spell-checks the visible region and only after an idle time.

It's really sweet, I'll give you that.  :-)

Also the fact that the whole visible area is spell-checked, and not only the places point has been to, makes it really stand apart `flyspell`.

And I see you're getting some well deserved praise on Reddit. :-)

**Comment by minad** [comment]
I am trying something new now. For full buffer check I always forcibly rechecked the whole buffer. I changed this now to only recheck the pending region. This is probably slightly more brittle. Therefore I went with the naive solution so far. You can try it at https://github.com/minad/jinx/commit/95fbf28c4dba6144ea562a55829dbc7b0f93e334. Please let me know if this changes anything for you for repeated checks. Also please check if you still get the same number of misspellings with this patch.

**Comment by gusbrs** [comment]
> I am trying something new now. For full buffer check I always forcibly rechecked the whole buffer. I changed this now to only recheck the pending region. This is probably slightly more brittle. Therefore I went with the naive solution so far. You can try it at [95fbf28](https://github.com/minad/jinx/commit/95fbf28c4dba6144ea562a55829dbc7b0f93e334). Please let me know if this changes anything for you for repeated checks. Also please check if you still get the same number of misspellings with this patch.

Thank you! I'll test and report back.

In the meantime, and for the record only, simply removing the call to `(seq-difference jinx--exclude-faces inhibit)` in  my `gb/jinx--latex-face-ignored-p` reduced the time from ~10s to ~4s. Given the "vanilla" time for that file of ~3s, it seems reasonable. But, go figure, a `seq-difference` on a short list...

**Comment by minad** [comment]
> Thank you! I'll test and report back.

Thanks.

> In the meantime, and for the record only, simply removing the call to (seq-difference jinx--exclude-faces inhibit) in my gb/jinx--latex-face-ignored-p reduced the time from ~10s to ~4s. Given the "vanilla" time for that file of ~3s, it seems reasonable. But, go figure, a seq-difference on a short list...

`seq-difference` is not a good idea. It is an inefficient and heavy function. This matters when called many times even on small lists. You should avoid any code which allocates in the predicates. The built-in Jinx predicates are optimized accordingly.

**Comment by gusbrs** [comment]
> `seq-difference` is not a good idea. It is an inefficient and heavy function. This starts to matter when called many times even on small lists. You should avoid any code which allocates in the predicates.

Learning by mistake. I had no idea it could cascade to have such an overall impact. I don't know how to recognize a code that allocates from one which doesn't, but I'll try to learn and keep that in mind. Thanks.

> When doing the aforementioned benchmarks, please disable any of your customizations. I would like to know if it changes anything given the default Jinx settings.

Well, ok, but I was using that just to diagnose my personal predicate, since we recognize the results of that call to be more than dubious. If we want to be able to take the benchmark results a little more seriously, we should probably improve them a little. I've tried to call `(benchmark-run 10 (progn (jit-lock-refontify) (jinx--force-check-region (point-min) (point-max))))`, but then I get a backtrace, which occurs occasionally anyway for that file, but this way it occurs every time and I can't get the benchmark to finish. Do you have any suggestions on a better benchmark call?

**Comment by minad** [comment]
Actually I don't need serious benchmark results. I just want to know if https://github.com/minad/jinx/commit/95fbf28c4dba6144ea562a55829dbc7b0f93e334 makes a perceivable for you for repeated checks in a very large file.

**Comment by gusbrs** [comment]
> Actually I don't need serious benchmark results. I just want to know if [95fbf28](https://github.com/minad/jinx/commit/95fbf28c4dba6144ea562a55829dbc7b0f93e334) makes a perceivable for you for repeated checks in a very large file. You said before that Jinx takes about 3s for your file. So if this would drop to half a second or so it would be interesting.

What I did, opened the file anew, with my customizations disabled. Called `(benchmark-run 1 (jinx--force-check-region (point-min) (point-max)))`, just once, because we want to capture the difference between the first call to the following ones. Result `(6.9441480140000005 36 2.8934784259999944)`. At this point `(length (jinx--get-overlays (point-min) (point-max)))` returns 11300. Called `(benchmark-run 1 (jinx--force-check-region (point-min) (point-max)))` again with result `(5.483370438 29 2.294937410000003)`. And again `(5.483370438 29 2.294937410000003)` returning the same 11300.

It seems ~3s presumes repeated calls and some of the distortions of not accounting for the repetition of font-locking. And that's also one difference between the first and the second call above which is not accounted for.

**Comment by minad** [comment]
Okay, I see. Not that of a large large difference with 5.5s vs 6.9s. I will probably keep the naive and safer implementation.

**Comment by gusbrs** [comment]
> Okay, I see. Not that of a large large difference with 5.5s vs 6.9s. I will probably keep the naive and safer implementation.

Yes, not a big difference. And makes sense.

Also, I said I had been getting some backtraces on this file. They are intermittent, so I haven't figured out the regularity, but repeating the force check I'm getting them somewhat frequently. Here is one such backtrace:

```
Debugger entered--Lisp error: (error "Error in syntax_table logic for to-the-end interva...")
  re-search-forward("\\<\\w+\\>" 1275557 t)
  (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start (match-beginning 0)) (word-end (match-end 0))) (while (and (< word-start word-end) (let ((c (char-after word-start))) (or (= c 39) (= c 8217)))) (setq word-start (1+ word-start))) (while (and (< word-start word-end) (let ((c (char-before word-end))) (or (= c 39) (= c 8217)))) (setq word-end (1- word-end))) (while (< word-start word-end) (let ((subword-end word-end)) (if jinx--camel (progn (goto-char word-start) (if (looking-at "\\([[:upper:]]?[[:lower:]]+\\)\\(?:[[:upper:]][[:lowe...") (progn ...)))) (goto-char subword-end) (let* ((val (run-hook-with-args-until-success ... word-start))) (cond ((integerp val) (let ... ...)) ((null val) (let nil ...)))) (setq word-start subword-end)))))
  (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start (match-beginning 0)) (word-end (match-end 0))) (while (and (< word-start word-end) (let ((c ...)) (or (= c 39) (= c 8217)))) (setq word-start (1+ word-start))) (while (and (< word-start word-end) (let ((c ...)) (or (= c 39) (= c 8217)))) (setq word-end (1- word-end))) (while (< word-start word-end) (let ((subword-end word-end)) (if jinx--camel (progn (goto-char word-start) (if ... ...))) (goto-char subword-end) (let* ((val ...)) (cond (... ...) (... ...))) (setq word-start subword-end))))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end 'jinx--pending t))))
  (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start (match-beginning 0)) (word-end (match-end 0))) (while (and (< word-start word-end) (let (...) (or ... ...))) (setq word-start (1+ word-start))) (while (and (< word-start word-end) (let (...) (or ... ...))) (setq word-end (1- word-end))) (while (< word-start word-end) (let ((subword-end word-end)) (if jinx--camel (progn ... ...)) (goto-char subword-end) (let* (...) (cond ... ...)) (setq word-start subword-end))))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end 'jinx--pending t)))) (set-match-data saved-match-data t))
  (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start (match-beginning 0)) (word-end (match-end 0))) (while (and (< word-start word-end) (let ... ...)) (setq word-start (1+ word-start))) (while (and (< word-start word-end) (let ... ...)) (setq word-end (1- word-end))) (while (< word-start word-end) (let (...) (if jinx--camel ...) (goto-char subword-end) (let* ... ...) (setq word-start subword-end))))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end 'jinx--pending t)))) (set-match-data saved-match-data t)))
  (save-excursion (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start ...) (word-end ...)) (while (and ... ...) (setq word-start ...)) (while (and ... ...) (setq word-end ...)) (while (< word-start word-end) (let ... ... ... ... ...)))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end 'jinx--pending t)))) (set-match-data saved-match-data t))))
  (progn (save-excursion (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let (... ...) (while ... ...) (while ... ...) (while ... ...))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end ... t)))) (set-match-data saved-match-data t)))))
  (unwind-protect (progn (save-excursion (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start ...)) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end ...)) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ... ... ... ...)) (remove-list-of-text-properties start end '...) (if retry-start (progn ...))) (set-match-data saved-match-data t))))) (if (or (not modified) (eq modified 'autosaved)) (progn (restore-buffer-modified-p modified))))
  (let* ((modified (buffer-modified-p)) (buffer-undo-list t) (inhibit-read-only t) (inhibit-modification-hooks t)) (unwind-protect (progn (save-excursion (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if ... nil ... ...) (goto-char end) (if ... nil ... ...) (jinx--delete-overlays start end) (goto-char start) (while ... ...) (remove-list-of-text-properties start end ...) (if retry-start ...)) (set-match-data saved-match-data t))))) (if (or (not modified) (eq modified 'autosaved)) (progn (restore-buffer-modified-p modified)))))
  (unwind-protect (let* ((modified (buffer-modified-p)) (buffer-undo-list t) (inhibit-read-only t) (inhibit-modification-hooks t)) (unwind-protect (progn (save-excursion (let ((saved-match-data ...)) (unwind-protect (progn ... ... ... ... ... ... ... ... ... ...) (set-match-data saved-match-data t))))) (if (or (not modified) (eq modified 'autosaved)) (progn (restore-buffer-modified-p modified))))) (set-syntax-table st))
  (let ((st (syntax-table)) case-fold-search retry-start retry-end) (unwind-protect (let* ((modified (buffer-modified-p)) (buffer-undo-list t) (inhibit-read-only t) (inhibit-modification-hooks t)) (unwind-protect (progn (save-excursion (let (...) (unwind-protect ... ...)))) (if (or (not modified) (eq modified 'autosaved)) (progn (restore-buffer-modified-p modified))))) (set-syntax-table st)))
  jinx--check-region(1 1275557)
  (closure ((end . 1275557) (start . 1)) nil (jinx--check-region start end))()
  funcall-with-delayed-message(1 "Checking..." (closure ((end . 1275557) (start . 1)) nil (jinx--check-region start end)))
  jinx--force-check-region(1 1275557)
  (progn (jinx--force-check-region (point-min) (point-max)) (or (jinx--get-overlays (point-min) (point-max)) (user-error "No misspellings in whole buffer")))
  (if all (progn (jinx--force-check-region (point-min) (point-max)) (or (jinx--get-overlays (point-min) (point-max)) (user-error "No misspellings in whole buffer"))) (or (jinx--get-overlays (window-start) (window-end) 'visible) (progn (jinx--force-check-region (window-start) (window-end)) (jinx--get-overlays (window-start) (window-end) 'visible)) (user-error "No misspelling in visible text")))
  (let* ((overlays (if all (progn (jinx--force-check-region (point-min) (point-max)) (or (jinx--get-overlays (point-min) (point-max)) (user-error "No misspellings in whole buffer"))) (or (jinx--get-overlays (window-start) (window-end) 'visible) (progn (jinx--force-check-region (window-start) (window-end)) (jinx--get-overlays (window-start) (window-end) 'visible)) (user-error "No misspelling in visible text")))) (count (length overlays)) (idx 0)) (if all (progn (push-mark))) (while (let* ((ov (and t (nth idx overlays)))) (if ov (let* ((deleted (not ...)) (skip (catch ... ...))) (cond ((integerp skip) (setq idx ...)) ((or all deleted) (setq idx ...)))) nil))))
  (unwind-protect (let* ((overlays (if all (progn (jinx--force-check-region (point-min) (point-max)) (or (jinx--get-overlays ... ...) (user-error "No misspellings in whole buffer"))) (or (jinx--get-overlays (window-start) (window-end) 'visible) (progn (jinx--force-check-region ... ...) (jinx--get-overlays ... ... ...)) (user-error "No misspelling in visible text")))) (count (length overlays)) (idx 0)) (if all (progn (push-mark))) (while (let* ((ov (and t (nth idx overlays)))) (if ov (let* ((deleted ...) (skip ...)) (cond (... ...) (... ...))) nil)))) (if old-point (progn (goto-char old-point))) (jit-lock-refontify))
  (progn (fset #'jinx--timer-handler #'ignore) (unwind-protect (let* ((overlays (if all (progn (jinx--force-check-region ... ...) (or ... ...)) (or (jinx--get-overlays ... ... ...) (progn ... ...) (user-error "No misspelling in visible text")))) (count (length overlays)) (idx 0)) (if all (progn (push-mark))) (while (let* ((ov (and t ...))) (if ov (let* (... ...) (cond ... ...)) nil)))) (if old-point (progn (goto-char old-point))) (jit-lock-refontify)))
  (unwind-protect (progn (fset #'jinx--timer-handler #'ignore) (unwind-protect (let* ((overlays (if all (progn ... ...) (or ... ... ...))) (count (length overlays)) (idx 0)) (if all (progn (push-mark))) (while (let* ((ov ...)) (if ov (let* ... ...) nil)))) (if old-point (progn (goto-char old-point))) (jit-lock-refontify))) (fset #'jinx--timer-handler old))
  (let* ((vnew (and (not all) (point-marker))) (old (symbol-function #'jinx--timer-handler)) (old-point vnew)) (unwind-protect (progn (fset #'jinx--timer-handler #'ignore) (unwind-protect (let* ((overlays (if all ... ...)) (count (length overlays)) (idx 0)) (if all (progn (push-mark))) (while (let* (...) (if ov ... nil)))) (if old-point (progn (goto-char old-point))) (jit-lock-refontify))) (fset #'jinx--timer-handler old)))
  jinx-correct((4))
  funcall-interactively(jinx-correct (4))
  command-execute(jinx-correct)
```
`1275557` is `(point-max)` on the file.

**Comment by minad** [comment]
Yes, I've also seen that rarely on Emacs 29. For example it happens in the aformentioned xterm.c (from the Emacs 29 source) if I repeatedly call `C-u M-x jinx-correct` to trigger a full buffer spell checking. I am getting stack traces like these:

~~~ elisp
Debugger entered--Lisp error: (error "Error in syntax_table logic for to-the-end interva...")
  (font-lock-fontify-keywords-region 1663 929795 nil)
  (font-lock-default-fontify-region 1663 929795 nil)
  (c-font-lock-fontify-region 1669 929795 nil)
  (font-lock-fontify-region 1669 929795)
  (#f(compiled-function (fun) #<bytecode -0x15557693fd12fe03>) font-lock-fontify-region)
  (jit-lock--run-functions 1669 929795)
  (jit-lock-fontify-now)
  ((closure (t) nil (jit-lock-fontify-now)))
  (funcall-with-delayed-message 1 "Fontifying..." (closure (t) nil (jit-lock-fontify-now)))
  (jinx--force-check-region 1 929795)
~~~

~~~ elisp
Debugger entered--Lisp error: (error "Error in syntax_table logic for to-the-end interva...")
  (re-search-forward "\\<\\w+\\>" 929795 t)
  (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start (match-beginning 0)) (word-end (match-end 0))) (while (and (< word-start word-end) (let ((c (char-after word-start))) (or (= c 39) (= c 8217)))) (setq word-start (1+ word-start))) (while (and (< word-start word-end) (let ((c (char-before word-end))) (or (= c 39) (= c 8217)))) (setq word-end (1- word-end))) (while (< word-start word-end) (let ((subword-end word-end)) (if jinx--camel (progn (goto-char word-start) (if (looking-at "\\([[:upper:]]?[[:lower:]]+\\)\\(?:[[:upper:]][[:lowe...") (progn ...)))) (goto-char subword-end) (let* ((val (run-hook-with-args-until-success ... word-start))) (cond ((integerp val) (let ... ...)) ((null val) (let nil ...)))) (setq word-start subword-end)))))
  (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start (match-beginning 0)) (word-end (match-end 0))) (while (and (< word-start word-end) (let ((c ...)) (or (= c 39) (= c 8217)))) (setq word-start (1+ word-start))) (while (and (< word-start word-end) (let ((c ...)) (or (= c 39) (= c 8217)))) (setq word-end (1- word-end))) (while (< word-start word-end) (let ((subword-end word-end)) (if jinx--camel (progn (goto-char word-start) (if ... ...))) (goto-char subword-end) (let* ((val ...)) (cond (... ...) (... ...))) (setq word-start subword-end))))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end 'jinx--pending t))))
  (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start (match-beginning 0)) (word-end (match-end 0))) (while (and (< word-start word-end) (let (...) (or ... ...))) (setq word-start (1+ word-start))) (while (and (< word-start word-end) (let (...) (or ... ...))) (setq word-end (1- word-end))) (while (< word-start word-end) (let ((subword-end word-end)) (if jinx--camel (progn ... ...)) (goto-char subword-end) (let* (...) (cond ... ...)) (setq word-start subword-end))))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end 'jinx--pending t)))) (set-match-data saved-match-data t))
  (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start (match-beginning 0)) (word-end (match-end 0))) (while (and (< word-start word-end) (let ... ...)) (setq word-start (1+ word-start))) (while (and (< word-start word-end) (let ... ...)) (setq word-end (1- word-end))) (while (< word-start word-end) (let (...) (if jinx--camel ...) (goto-char subword-end) (let* ... ...) (setq word-start subword-end))))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end 'jinx--pending t)))) (set-match-data saved-match-data t)))
  (save-excursion (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ((word-start ...) (word-end ...)) (while (and ... ...) (setq word-start ...)) (while (and ... ...) (setq word-end ...)) (while (< word-start word-end) (let ... ... ... ... ...)))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end 'jinx--pending t)))) (set-match-data saved-match-data t))))
  (progn (save-excursion (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start (match-beginning 0))) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end (match-beginning 0))) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let (... ...) (while ... ...) (while ... ...) (while ... ...))) (remove-list-of-text-properties start end '(jinx--pending)) (if retry-start (progn (put-text-property retry-start retry-end ... t)))) (set-match-data saved-match-data t)))))
  (unwind-protect (progn (save-excursion (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if (looking-at-p "\\<") nil (re-search-backward "\\<\\|^") (setq start ...)) (goto-char end) (if (looking-at-p "\\>") nil (re-search-forward "\\>\\|$") (setq end ...)) (jinx--delete-overlays start end) (goto-char start) (while (re-search-forward "\\<\\w+\\>" end t) (let ... ... ... ...)) (remove-list-of-text-properties start end '...) (if retry-start (progn ...))) (set-match-data saved-match-data t))))) (if (or (not modified) (eq modified 'autosaved)) (progn (restore-buffer-modified-p modified))))
  (let* ((modified (buffer-modified-p)) (buffer-undo-list t) (inhibit-read-only t) (inhibit-modification-hooks t)) (unwind-protect (progn (save-excursion (let ((saved-match-data (match-data))) (unwind-protect (progn (set-syntax-table jinx--syntax-table) (goto-char start) (if ... nil ... ...) (goto-char end) (if ... nil ... ...) (jinx--delete-overlays start end) (goto-char start) (while ... ...) (remove-list-of-text-properties start end ...) (if retry-start ...)) (set-match-data saved-match-data t))))) (if (or (not modified) (eq modified 'autosaved)) (progn (restore-buffer-modified-p modified)))))
  (unwind-protect (let* ((modified (buffer-modified-p)) (buffer-undo-list t) (inhibit-read-only t) (inhibit-modification-hooks t)) (unwind-protect (progn (save-excursion (let ((saved-match-data ...)) (unwind-protect (progn ... ... ... ... ... ... ... ... ... ...) (set-match-data saved-match-data t))))) (if (or (not modified) (eq modified 'autosaved)) (progn (restore-buffer-modified-p modified))))) (set-syntax-table st))
  (let ((st (syntax-table)) case-fold-search retry-start retry-end) (unwind-protect (let* ((modified (buffer-modified-p)) (buffer-undo-list t) (inhibit-read-only t) (inhibit-modification-hooks t)) (unwind-protect (progn (save-excursion (let (...) (unwind-protect ... ...)))) (if (or (not modified) (eq modified 'autosaved)) (progn (restore-buffer-modified-p modified))))) (set-syntax-table st)))
  (jinx--check-region 1 929795)
~~~

This seems like an Emacs bug. Some kind of internal assertion violation. We should probably report this via `report-emacs-bug`. I've only found a question on [Stackoverflow](https://stackoverflow.com/questions/21189602/error-in-syntax-table-logic-for-to-the-end-intervals-error-in-emacs), where a user called "Stefan" also declared this an Emacs bug. I boldly assume that this is the former Emacs maintainer Stefan @monnier, so let's just ping him. :)

**Comment by gusbrs** [comment]
> Yes, I've also seen that rarely on Emacs 29. For example it happens in the aformentioned xterm.c (from the Emacs 29 source) if I repeatedly call `C-u M-x jinx-correct` to trigger a full buffer spell checking.

Yup. From the Messages buffer here:

```
Fontifying...
font-latex-match-command-with-arguments: Error in syntax_table logic for to-the-end intervals
Fontifying...
font-latex-match-command-with-arguments: Error in syntax_table logic for to-the-end intervals
Fontifying...
Checking...
looking-at-p: Error in syntax_table logic for to-the-end intervals
```

> This seems like an Emacs bug. Some kind of internal assertion violation. We should probably report this via `report-emacs-bug`. I've only found a question on [Stackoverflow](https://stackoverflow.com/questions/21189602/error-in-syntax-table-logic-for-to-the-end-intervals-error-in-emacs), where a user called "Stefan" also declared this an Emacs bug. I boldly assume that this is the former Emacs maintainer Stefan monnier, so let's just ping him. :)

I'm not sure what to report at this point. The intermittent ones are the worst. Let's hope your bold ping works. :)

**Comment by minad** [comment]
@gusbrs Interesting. Can you create a minimal recipe starting from emacs -Q and share a file where this happens reproducibly? I've so far only seen this in my Emacs configuration.

**Comment by gusbrs** [comment]
Btw, even if you've decided to keep the naive and safer implementation for the checks, are you still considering what you mentioned above?

> Regarding your second point I think I could refactor the code somewhat such that the list of checked overlays is more flexible or could be passed as argument even. This would make the checking code more easily tweakable for the user.

That would be nice...

**Comment by minad** [comment]
> Btw, even if you've decided to keep the naive and safer implementation for the checks, are you still considering what you mentioned above?

It is not a priority.

**Comment by gusbrs** [comment]
> It is not a priority.

Understood. And thanks for considering.

**Comment by monnier** [comment]
> Well, me too, not to mention the "complex large file". And, as I said,
> I have no idea what triggers it. So far, the only regular way to generate
> a backtrace here is `(benchmark-run 10 (progn (jit-lock-refontify)
> (jinx--force-check-region (point-min) (point-max))))`. On that file, with my
> config (minus my Jinx setup for LaTeX, as agreed).

The error you post is an internal error in Emacs's C code.   If you can
reproduce it reliably with files you can share (at least with me), then
please `M-x report-emacs-bug` and post the recipe (starting with
`emacs -Q`) there (and if some of the files are too sensitive to be made
public send them to me privately).

Sweet&short recipes and files are preferable, of course, but
not indispensable.

**Comment by gusbrs** [comment]
> The error you post is an internal error in Emacs's C code. If you can reproduce it reliably with files you can share (at least with me), then please `M-x report-emacs-bug` and post the recipe (starting with `emacs -Q`) there (and if some of the files are too sensitive to be made public send them to me privately).
>
> Sweet&short recipes and files are preferable, of course, but not indispensable.

I can try and, if I succeed, I can report it. But, would it make sense to make a bug report to Emacs using `jinx`? Also I'm afraid this particular file is not shareable. Anyway, let's see what I can find.

**Comment by gusbrs** [comment]
@monnier and @minad , sorry, but I can't seem to be able to isolate the problem from my config. Calls to `(benchmark-run 10 (progn (jit-lock-refontify) (jinx--force-check-region (point-min) (point-max))))` reliably generate a backtrace to this error either in my personal tex file or in `xterm.c` (at least independent of my file). But I couldn't trim down further than that. No errors with `emacs -Q`. On the other hand, I have zero `c-mode` config, which means the culprit must be something more mundane.

@minad, since you also occasionally get backtraces, can you check if you can reliably get them with that call on `xterm.c`?

**Comment by minad** [comment]
@monnier @gusbrs Okay, I found a "minimal" recipe. It is weird. It has something to do with `inhibit-message` and I also used Jinx which temporarily modifies the syntax table from a timer (not sure if Jinx is doing something wrong there, at least not obviously). Can you confirm that you see this issue with the given recipe?

~~~ elisp
;; 1. Evaluate code

(setq debug-on-error t)
(setq inhibit-message-regexps (list "\\`bug"))
(add-to-list 'set-message-functions #'inhibit-message)
(package-initialize)
(global-jinx-mode)

;; 2. Open xdisp.c
;; 3. C-u M-x jinx-correct

;; Debugger entered--Lisp error: (error "Error in syntax_table logic for to-the-end interva...")
;;   font-lock-fontify-keywords-region(4479 1215610 nil)
;;   font-lock-default-fontify-region(4479 1215610 nil)
;;   c-font-lock-fontify-region(4501 1215610 nil)
;;   font-lock-fontify-region(4501 1215610)
;;   #f(compiled-function (fun) #<bytecode -0x1551804fd6283e03>)(font-lock-fontify-region)
;;   jit-lock--run-functions(4501 1215610)
;;   jit-lock-fontify-now()
;;   #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_8>()
;;   jinx--force-check-region(1 1215610)
;;   jinx-correct((4))
~~~

**Comment by minad** [comment]
@gusbrs Can you please check if you still observe the issue even without https://github.com/minad/jinx/commit/ef5dd1625cc3ddd7c2345c9ce49d7f82be681c3b. If my suspicion is confirmed I can try to create a minimal recipe involving only `with-delayed-message` and `inhibit-message` and report this upstream.

**Comment by minad** [comment]
@gusbrs Regarding multi language support I had the idea that we could add a simple heuristic to detect the language of the current sentence by looking at the last three to five words and testing against the available dictionaries. This would help avoiding false negatives in mixed language documents as you mentioned initially in the issue.

**Comment by gusbrs** [comment]
> Can you please check if you still observe the issue even without [ef5dd16](https://github.com/minad/jinx/commit/ef5dd1625cc3ddd7c2345c9ce49d7f82be681c3b)? If my suspicion is confirmed I can try to create a minimal recipe involving only `with-delayed-message` and `inhibit-message` and report this upstream.

Yes, I can confirm the issue is gone on my .tex file after you removed the calls to `with-delayed-message`. I can also call `C-u M-x jinx-correct` on `xterm.c`. But I can still get a backtrace for `(benchmark-run 10 (jinx--force-check-region (point-min) (point-max)))` on it, albeit a different one.

```lisp
Debugger entered--Lisp error: (wrong-type-argument syntax-table-p nil)
  jinx--check-region(1 929795)
  jinx--force-check-region(1 929795)
  (closure (t) nil (jinx--force-check-region (point-min) (point-max)))()
  benchmark-call((closure (t) nil (jinx--force-check-region (point-min) (point-max))) 10)
  eval-expression((benchmark-run 10 (jinx--force-check-region (point-min) (point-max))) nil nil 127)
  funcall-interactively(eval-expression (benchmark-run 10 (jinx--force-check-region (point-min) (point-max))) nil nil 127)
  command-execute(eval-expression)
```
Thanks for the provisional workaround, btw.

> Regarding multi language support I had the idea that we could add a simple heuristic to detect the language of the current sentence by looking at the last three to five words and testing against the available dictionaries. This would help avoiding false negatives in mixed language documents as you mentioned initially in the issue.

Some sort of support for this would be great. Thank you very much for looking into this! I don't know how an heuristic like that would perform in terms of speed and results. I don't really recall how well `guess-language` did in this area, but I was more troubled by the constant changing of the global dictionary, which would not happen with Jinx. So I'm looking forward to it.

Now, what would make me really, really happy is if you exposed some sort of `jinx-current-dict-function`, returning the dictionary to use when checking or correcting. Then I could apply font-locking of LaTeX documents (like done for fine tuning the check rules for arguments of macros at https://github.com/minad/jinx/issues/25#issuecomment-1511954696), now that I learned the basics of that machinery. It is a pretty powerful approach. I think you'll be interested in knowing that now I can add `font-lock-constant-face` for my `tex-mode` `jinx-exclude-faces`, handling in one fell swoop all bibkeys in citation commands and labels in reference commands. But then I can add back and not ignore the few problematic ones of those classes, like `\footnote`, `\footnotetext` and `\marginpar`.  ;-)

Btw, I eventually chimed in into that thread at emacs-devel about it (https://lists.gnu.org/archive/html/emacs-devel/2023-04/msg00512.html, and a related report with implications for this issue https://lists.gnu.org/archive/html/auctex-devel/2023-04/msg00006.html). Let's see if Arash, or someone else, gets interested. If support to adding this needed extra information is done directly to the font-lock machinery it would be ideal.

**Comment by minad** [comment]
@gusbrs That's unfortunate, since this means that there is a deeper bug somewhere in Emacs.

> I don't really recall how well guess-language did in this area, but I was more troubled by the constant changing of the global dictionary, which would not happen with Jinx. So I'm looking forward to it.

It doesn't make much sense to use guess language here since we have access to the dictionaries directly and we could just ask them.

> Now, what would make me really, really happy is if you exposed some sort of jinx-current-dict-function, returning the dictionary to use when checking or correcting.

Yes, I had that idea at some point since it would allow multi-language documents where we could use different dictionaries per section. I hope that such a heuristic may be better than exposing such a function since the only examples for multi-language documents I've seen were not convincing (xml and TeX babel). You could try to locally modify `jinx--dicts` in the predicate.

**Comment by gusbrs** [comment]
> It doesn't make much sense to use guess language here since we have access to the dictionaries directly and we could just ask them.

> Yes, I had that idea at some point since it would allow multi-language documents where we could use different dictionaries per section.

I'm not sure I can envisage what you have in mind. But I'm looking forward to what you come up with.  :)

> I hope that such a heuristic may be better than exposing such a function since the only examples for multi-language documents I've seen were not convincing (xml and TeX babel).

They don't need to be mutually exclusive, do they? Also, other major-modes could leverage such API if it exists with some dedicated markup. E.g. an Org property for the heading, which you came up with yourself.

Besides, if you do come up with this heuristic approach, it will likely (and hopefully) be made optional in some form. Why not by means of the function which does the task?

> You could try to locally modify `jinx--dicts` in the predicate.

That might well work for checking but, as far as I can tell, `jinx--correct-suggestions` would still offer the full set when correcting.

**Comment by minad** [comment]
On 4/21/23 14:58, Gustavo Barros wrote:
> Besides, if you do come up with this heuristic approach, it will likely
> (and hopefully) be made optional in some form. Why not by means of the
> function which does the task?

Yes, it is a good idea to offer such a configurable dictionary function
which could be either set to a function just returning all dictionaries,
or returns dictionaries based on a heuristic or based on markup blocks.
But it is also not a high priority for me, just an idea so far.

**Comment by minad** [comment]
For reference, I submitted a bug report regarding the "Error in syntax_table logic for to-the-end intervals": https://debbugs.gnu.org/cgi/bugreport.cgi?bug=63253.
